swim/heal_partition_test.go

```diff
@@ -0,0 +1,344 @@
+	assert.Len(t, targets, 1, "expected that b healed with node node a")
+	assert.Equal(t, a.node.Address(), targets[0], "expected b healed with node a")
+
+	// a is now part of b's membership
+	_, has = b.node.memberlist.Member(a.node.Address())
+	assert.True(t, has, "expected that a is now part of b's membership")
+
+	// make sure that a is still not bootstrapped, this causes a to be suspect to be
+	ExecuteThenWaitFor(func() {
+		b.node.pingNextMember()
+	}, a.node, RequestBeforeReadyEvent{})
+
+	// allows a to bootstrap
+	ExecuteThenWaitFor(func() {
+		close(block)
+	}, a.node, JoinCompleteEvent{})
+
+	// progress timers so that incarnation numbers can bump
+	c := clock.NewMock()
+	c.Add(time.Millisecond)
+	a.node.clock = c
+
+	waitForConvergence(t, time.Second*3, a, b)
+}
+
+type partition []*testNode
+
+func newPartition(t *testing.T, n int) partition {
+	p := partition(genChannelNodes(t, n))
+	bootstrapNodes(t, p...)
+	waitForConvergence(t, 500*time.Millisecond, p...)
+	return p
+}
+
+func (p partition) AddPartitionWithStatus(B partition, status string) {
+	for _, n := range p {
+		for _, n2 := range B {
+			n.node.memberlist.MakeChange(n2.node.Address(), n2.node.Incarnation(), status)
+		}
+	}
+
+	for _, n := range p {
+		n.node.disseminator.ClearChanges()
+	}
+}
+
+func (p partition) ProgressTime(T time.Duration) {
+	for _, n := range p {
+		now := n.node.clock.Now()
+		c := clock.NewMock()
+		c.Add(now.Add(T).Sub(clock.NewMock().Now()))
+		n.node.clock = c
+	}
+}
+
+// HasPartitionAs checks that every node from A, contains every node from B in its
+// membership. It also checks that the members from B have the correct
+// incarnation number and status.
+func (p partition) HasPartitionAs(t *testing.T, B partition, incarnation int64, status string) {
+	for _, a := range p {
+		for _, b := range B {
+			mem, ok := a.node.memberlist.Member(b.node.Address())
+			assert.True(t, ok, "expected members to contain member")
+			assert.Equal(t, incarnation, mem.Incarnation, "expected correct incarnation number")
+			assert.Equal(t, status, mem.Status, "expected correct status")
+		}
+	}
+}
+
+func (p partition) DoesNotContain(t *testing.T, B partition) {
+	for _, a := range p {
+		for _, b := range B {
+			_, ok := a.node.memberlist.Member(b.node.Address())
+			assert.False(t, ok, "expected memberlist to not contain member")
+		}
+	}
+}
+
+func (p partition) Hosts() []string {
+	var res []string
+	for _, a := range p {
+		res = append(res, a.node.Address())
+	}
+	return res
+}
+
+func (p partition) String() string {
+	r := ""
+	mems := p[0].node.memberlist.GetMembers()
+	for i := range mems {
+		r += fmt.Sprintln(mems[i].Address, mems[i].Incarnation, mems[i].Status)
+	}
+	r += fmt.Sprintln()
+	return r
+}
+
+func (p partition) Contains(address string) bool {
+	for _, a := range p {
+		if a.node.Address() == address {
+			return true
+		}
+	}
+	return false
+}
+
+// waitForPartitionHeal is same as waitForConvergence but unlike that function,
```

Comment:
You should describe this function without referring how it's similar to other functions. :-) If it really makes sense that way, please copy (parts of) the godoc string.